LibCore + Utilities: Add first version of sed utility

rtobar opened the pull request:
This is a first version of the `sed` utility as described in https://pubs.opengroup.org/onlinepubs/9699919799/utilities/sed.html.

There is still some work to do, specially around supporting REs and such, but most of the basic functions and addressing modes are supported. I've been comparing outputs against GNU's sed to make sure things are correctly implemented.

The (small) change in LibCore is to make `LineIterator` movable, which allows me to keep it as a member in a structure that in turn I put into a Vector.

At some pint I'd really love to add tests for this, but I'll need some guidance on where to put those.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+
+    AddressType address_type() const { return m_address_type; }
+
+    bool matches(StringView /*pattern_space*/, size_t line_number, bool is_last_line) const
```
Let's use `[[maybe_unused]] StringView pattern_space`, the C-style `/*foo*/` is quite irritating to see.

rtobar replied:
True, I hadn't considered `[[maybe_unused]]`, fixed now.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+            builder.append('$');
+            break;
+        case AddressType::ContextAddress:
+            VERIFY_NOT_REACHED();
```
Is this a `TODO()`?

rtobar replied:
Mmm... it's both probably? This is the formatting routine for `Address`, but the execution can't reach this code at the moment because addresses of type `ContextAddress` are not yet parsed, so an `Address` with `ContextAddress` type is never constructed.

I'd say the TODO, if anywhere, would be more on the `parse_address` function, but if you think it's better to go with a `TODO()` here too I'm happy to change it.

alimpfard replied:
Either is fine, I just thought it might have been forgotten or something

rtobar replied:
I'll leave the `VERIFY_NOT_REACHED` for the time being then.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+}
+
+struct Command {
+    Address addr1;
+    Address addr2;
```
let's use full variable names.
```suggestion
    Address address1;
    Address address2;
```

rtobar replied:
Done

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    bool is_enabled() const { return m_is_enabled; }
+
+private:
+    bool m_is_enabled { false };
+    bool m_is_selection_active { false };
```
Please put these at the end of the struct definition (after the `selects()` member function).

rtobar replied:
Done

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    }
+
+    size_t index = 0;
+    for (; index < sv.length() && '0' <= sv[index] && sv[index] <= '9'; index++)
+        ;
+    if (index == 0)
+        return { {}, 0 };
+    return {
+        Address { AK::StringUtils::convert_to_uint<size_t>(sv.substring_view(0, index)).release_value() },
+        index
+    };
```
I was thinking that this looked quite similar to how `strtoull` would work, but that doesn't take ptr+length - sad :(

rtobar replied:
Yeap, this is the best I could come up with in terms of reusing code after exploring AK for a bit. I hadn't actually thought on using `strtoull` actually, but as you say it's not a viable option anyways.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+}
+
+template<int max_addresses>
+static bool verify_number_of_addresses(const Command& command)
```
Please use east-const in new files.

rtobar replied:
Fixed. I would have expected clang-format to have changed that for me, I thought it did in other occasions?

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+template<int max_addresses>
+static bool verify_number_of_addresses(const Command& command)
+{
+    if constexpr (max_addresses == 2)
```
As the `else` block has braces, please put braces around the then block too.

rtobar replied:
Done

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    size_t last_index;
+};
+
+static CommandParsingResult parse_command(StringView command_p)
```
This parameter is copied in the body and never used later, perhaps this should just be
```suggestion
static CommandParsingResult parse_command(StringView to_parse)
```

rtobar replied:
Ahh, a leftover from the first iterations of the code, thanks for spotting it, fixed now

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    char command_char = to_parse[0];
+    to_parse = to_parse.substring_view(1);
+    end_of_command_index += 1;
+#define HANDLE_FUNCTION_CASE(c, max_addresses)                     \
```
Let's leave a bit of whitespace above this, it looks a little too crowded here :P

rtobar replied:
Added, it was unreadable indeed

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+        if (!verify_number_of_addresses<max_addresses>(command)) { \
+            return { {}, end_of_command_index };                   \
+        }                                                          \
+        break;
```
Please leave an empty line after this, I thought this was a switch-in-a-switch scenario for a second

rtobar replied:
Added

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    bool m_started { false };
+};
+
+static void write_pattern_space(RefPtr<Core::File> output, StringBuilder& pattern_space)
```
Can this take a `Core::File&` instead? no need to go through a ref/unref just to treat it as a simple reference.

rtobar replied:
Sure can, fixed.now

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    output->write("\n");
+}
+
+static void print_unambiguous(RefPtr<Core::File> output, StringView pattern_space)
```
Same note about `Core::File&` here.

rtobar replied:
Done

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+
+        if (cycle_decision == CycleDecision::Next) {
+            continue;
+        } else if (cycle_decision == CycleDecision::Quit) {
```
No need for `else` after that `continue`.

rtobar replied:
I don't understand the comment. `cycle_decision` can be `None`, `Next` or `Quit`, and in this `if`/`else if` block the three values give the three different, expected behaviors. Not having this `else if` here wouldn't allow me to implement the `Quit` cycle decision.

alimpfard replied:
I was suggesting this:
```suggestion
        if (cycle_decision == CycleDecision::Quit) {
```

(i.e. just the "else", not the whole else-if)

rtobar replied:
Ahhh, I see now, sorry for the confusion, changed now.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+        .accept_value = [&script](const char* script_file) {
+            auto file = Core::File::open(script_file, Core::OpenMode::ReadOnly);
+            if (file.is_error()) {
+                perror("Failed to open script file");
```
I wouldn't trust `perror` to tell you the right error here, `file.error()` already has the correct error string anyway:
```suggestion
                warnln("Failed to open script file: {}", file.error());
```

rtobar replied:
Thanks for pointing out, I think I'm still not terribly familiar with the Serenity libs and their intended usage. Fixed now.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    });
+    arg_parser.add_positional_argument(pos_args, "script and/or file", "...", Core::ArgsParser::Required::No);
+    if (!arg_parser.parse(argc, argv)) {
+        VERIFY_NOT_REACHED();
```
This won't happen, if `parse()` fails, it'll just `exit(1)`.
you don't need to check the return value unless you pass `FailureBehavior::Ignore` or `FailureBehavior::PrintUsage`.

rtobar replied:
Simplified now, thanks for pointing this out.

alimpfard commented on Userland/Utilities/sed.cpp:
```diff
@@ -0,0 +1,559 @@
+    for (const auto& filename : pos_args) {
+        auto file = Core::File::open(filename, Core::OpenMode::ReadOnly);
+        if (file.is_error()) {
+            perror("Cannot open input file for reading");
```
Same note about `perror` and `File::open()` here.

rtobar replied:
Fixed.

rtobar commented:
It's been a while since I had some time to work again on this. Updates on latest push:

 * Rebased this onto the latest master
 * Updated `Formatter` classes for new return type

Otherwise I believe all original change requests have been addressed.

bgianfo commented:
> At some pint I'd really love to add tests for this, but I'll need some guidance on where to put those.

Some directory under `Tests/@`? And hook them up so they get run by `run-tests`. So you want to write she'll script wrappers which call / validate the results? Or do you want to make sed into a library that you can write `LibTest` based unit tests around?

rtobar commented:
> > At some pint I'd really love to add tests for this, but I'll need some guidance on where to put those.
> 
> Some directory under `Tests/@`? And hook them up so they get run by `run-tests`. So you want to write she'll script wrappers which call / validate the results? Or do you want to make sed into a library that you can write `LibTest` based unit tests around?

Neither. I had the idea of extending `Core::Command` to accept input as well,  so I could write unit tests launching sed with different arguments and standard input and I could check the output. Writing shell scripts should work, but I wanted to avoid the dependency on the shell (also the tests would have to go under Tests/Shell? Not sure). Another possibility is to skip this step and always tests sed against input files, not the stdin, which should be doable.

I did start some Tests/Utilities subdir with small test for echo, which doesn't require stdin, but haven't gotten around to PRing them yet.
